pkg/ipam-node/store/store.go

```diff
@@ -0,0 +1,315 @@
+	defer s.lock.Unlock()
+}
+
+func (s *store) markClosed() {
+	s.isClosed = true
+}
+
+func (s *store) checkClosed() {
+	if s.isClosed {
+		panic("operation was performed on the closed store")
+	}
+}
+
+// Reserve is the Store interface implementation for store
+func (s *store) Reserve(pool string, id string, ifName string, meta types.ReservationMetadata, address net.IP) error {
+	s.checkClosed()
+	reservationKey := s.getKey(id, ifName)
+	poolData := s.getPoolData(pool, s.tmpData)
+	_, exist := poolData.Entries[reservationKey]
+	if exist {
+		return ErrReservationAlreadyExist
+	}
+	duplicateIP := false
+	for _, r := range poolData.Entries {
+		if address.Equal(r.IPAddress) {
+			duplicateIP = true
+			break
+		}
+	}
+	if duplicateIP {
+		// duplicate allocator should retry
+		return ErrIPAlreadyReserved
+	}
+	poolData.LastReservedIP = address
+	poolData.LastPoolConfig = meta.PoolConfigSnapshot
+	reservation := types.Reservation{
+		ContainerID:   id,
+		InterfaceName: ifName,
+		IPAddress:     address,
+		Metadata:      meta,
+	}
+	reservation.Metadata.CreateTime = time.Now().Format(time.RFC3339Nano)
+	poolData.Entries[reservationKey] = reservation
+	s.tmpData.Pools[pool] = *poolData
+	s.isModified = true
+	return nil
+}
+
+func (s *store) getPoolData(pool string, layout *types.Root) *types.PoolReservations {
+	res, exist := layout.Pools[pool]
+	if exist {
+		if res.Entries == nil {
+			res.Entries = map[string]types.Reservation{}
+		}
+		return &res
+	}
+	return types.NewPoolReservations(pool)
+}
+
+// ListReservations is the Store interface implementation for store
+func (s *store) ListReservations(pool string) []types.Reservation {
+	s.checkClosed()
+	poolData := s.getPoolData(pool, s.tmpData)
+	allocations := make([]types.Reservation, 0, len(poolData.Entries))
+	for _, a := range poolData.Entries {
+		allocations = append(allocations, a)
+	}
+	return allocations
+}
+
+// ListPools is the Store interface implementation for store
+func (s *store) ListPools() []string {
+	s.checkClosed()
+	pools := make([]string, 0, len(s.tmpData.Pools))
+	for poolName := range s.tmpData.Pools {
+		pools = append(pools, poolName)
+	}
+	return pools
+}
+
+// GetLastReservedIP is the Store interface implementation for store
+func (s *store) GetLastReservedIP(pool string) net.IP {
+	s.checkClosed()
+	poolData := s.getPoolData(pool, s.tmpData)
+	return poolData.LastReservedIP
+}
+
+// SetLastReservedIP is the Store interface implementation for store
+func (s *store) SetLastReservedIP(pool string, ip net.IP) {
+	s.checkClosed()
+	poolData := s.getPoolData(pool, s.tmpData)
+	poolData.LastReservedIP = ip
+	s.tmpData.Pools[pool] = *poolData
+	s.isModified = true
+}
+
+// ReleaseReservationByID is the Store interface implementation for store
+func (s *store) ReleaseReservationByID(pool string, id string, ifName string) {
+	s.checkClosed()
+	poolData := s.getPoolData(pool, s.tmpData)
+	delete(poolData.Entries, s.getKey(id, ifName))
+	s.tmpData.Pools[pool] = *poolData
```

Comment:
when are we deleting a pool entry from store ? 

i dont think its being done today.

option 1:
remove a pool from store if it has no reservations here

option 2:
add API to remove pool and call it in cleaner if a pool has no reservations.